lib/spack/spack/binary_distribution.py

```diff
@@ -0,0 +1,409 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+# "Benedikt Hegner (CERN)"
+# "Patrick Gartung (FNAL)"
+
+import os
+import re
+import tarfile
+import yaml
+import shutil
+
+import llnl.util.tty as tty
+from spack.util.gpg import Gpg
+from llnl.util.filesystem import mkdirp, join_path
+from spack.util.web import spider
+import spack.cmd
+import spack
+from spack.stage import Stage
+import spack.fetch_strategy as fs
+import spack.relocate
+from contextlib import closing
+import spack.util.gpg as gpg_util
+
+
+def has_gnupg2():
+    try:
+        gpg_util.Gpg.gpg()('--version', output=os.devnull)
+        return True
+    except Exception:
+        return False
+
+
+def buildinfo_file_name(spec):
+    """
+    Filename of the binary package meta-data file
+    """
+    installpath = join_path(spack.store.layout.root,
+                            install_directory_name(spec))
+    return os.path.join(installpath, ".spack", "binary_distribution")
+
+
+def read_buildinfo_file(spec):
+    """
+    Read buildinfo file
+    """
+    filename = buildinfo_file_name(spec)
+    with open(filename, 'r') as inputfile:
+        content = inputfile.read()
+        buildinfo = yaml.load(content)
+    return buildinfo
+
+
+def write_buildinfo_file(spec):
+    """
+    Create a cache file containing information
+    required for the relocation
+    """
+    text_to_relocate = []
+    binary_to_relocate = []
+    blacklist = (".spack", "man")
+    for root, dirs, files in os.walk(spec.prefix, topdown=True):
+        dirs[:] = [d for d in dirs if d not in blacklist]
+        for filename in files:
+            path_name = os.path.join(root, filename)
+            filetype = spack.relocate.get_filetype(path_name)
+            if spack.relocate.needs_binary_relocation(filetype):
+                rel_path_name = os.path.relpath(path_name, spec.prefix)
+                binary_to_relocate.append(rel_path_name)
+            elif spack.relocate.needs_text_relocation(filetype):
+                rel_path_name = os.path.relpath(path_name, spec.prefix)
+                text_to_relocate.append(rel_path_name)
+
+    # Create buildinfo data and write it to disk
+    buildinfo = {}
+    buildinfo['buildpath'] = spack.store.layout.root
+    buildinfo['relocate_textfiles'] = text_to_relocate
+    buildinfo['relocate_binaries'] = binary_to_relocate
+    filename = buildinfo_file_name(spec)
+    with open(filename, 'w') as outfile:
+        outfile.write(yaml.dump(buildinfo, default_flow_style=True))
+
+
+def install_directory_name(spec):
```

Comment:
It needs to be spec. See line 113: I use spec.name, spec.version, spec.dag_hash().